src/ActionInterface.cpp

```diff
@@ -366,6 +366,13 @@ void ActionInterface::OffsetStandbyFrequency(double offset_khz, bool to_devices)
 
 void ActionInterface::ExchangeRadioFrequencies(bool to_devices)
 {
+  /* send to external devices */
+  if (to_devices) {
+    MessageOperationEnvironment env;
+    if (device_blackboard->ExchangeRadioFrequencies(env))
+      return;
```

Comment:
Has this failed due to an error or because the driver doesn't support it?
If there was an error, there should not be a fallback.

Reply:
It returns if ExchangeRadioFrequencies was successful, thus not executing the fallback. I could reverse the statement so the fallback gets called directly on unsuccessful frequency exchange.

Reply:
Maybe I misunderstood your question.
The previous implementation failed due to one of the drivers (and its device) does not support a direct setting of the active frequency, but it supports an exchange frequency command (1ac47d1a3c).
Because of our previous discussion if XCSoar should support a fallback or not, I've implemented a way to directly exchange the frequencies in the driver and fallback if the method is not implemented by it (or fails).

Reply:
I have no idea what you are talking about, but none of this answers my question.
Your code will never use the fallback code.
Did you test this? Did it work? Please demonstrate that I'm wrong.

Reply:
Yes, I tested it and it works for me. I'll have a look into your comments to investigate the flaws.

I thought about this again in general and I don't see (or maybe understand) the usefulness of the fallback at all. All devices I know and checked have different capabilities. Some have allow exchange commands, some allow setting one frequency, some both, some allow activating dual watch. The best place to cope with the commands and respective error handling is in the drivers themselves. Its another thing of course for stuff like QNH or baro altitude, which XCSoar needs as well for internal calculations, but XCSoar has no use for radio frequencies without an attached device.

Reply:
> The best place to cope with the commands and respective error handling is in the drivers themselves.

That is not what your code submission does. Your code submission breaks the feature for existing drivers.

Reply:
That is correct. Unfortunately, you cherry-picking parts of the PR broke it for one of the devices as well.

I see two options:
1. create a fall-back in XCSoar to handle the feature according to the devices capabilities (i'm not content with that approach).
2. get all drivers to implement ExchangeRadioFrequencies themselves. I either need the original developers for that or all four devices that drivers exist for to test the implementation.

Reply:
Fallback in XCSoar core code is fine; I don't think it's a good idea to duplicate code in each driver.

Reply:
The drivers wouldn't duplicate code. Each would implement the exchange of frequencies in a way that is supported by its device, same way setting frequencies is handled by the device drivers.
Unfortunately I cannot check what the other radio devices are capable of because I cannot find any documentation on the protocols.

Reply:
Sigh, now you understand my strict "merge driver code only if vendor publishes documentation for free" policy which I'm often criticized for... disappearing documentation is obviously another problem. There were vendors who used copyright to prevent us to mirror their documentation. 10 years later, this is still sad.

So, if you take care for implementing this in all drivers, then yes I'd agree to your idea moving everything into the drivers.